### ISSUE TITLE
Macros inside generic class are executed without type params values

### ISSUE THREAD

**Issue by bcardiff** [crystal-lang/crystal#1445]
http://play.crystal-lang.org/#/r/eqk

macros inside generic class, but outside body are not expanded per instantiated type.
I would expect to do some macro expansion depending on type args.

```crystal
class Foo(S)
  {{ puts @type.type_params }}

  def foo
    {{ puts @type.type_params }}
  end
end

class P
end

Foo(P).new.foo
```

outputs 
```
[]
[P]
```

instead of
```
[P]
[P]
```

**Comment by waterlink** [comment]
Looks to me as if the body gets fully "compile-time-evaluated" even before
any concrete class gets defined..

--
Best Regards,
Alexey Fedorov,
Sr Ruby, Clojure, Crystal, Golang Developer,
Microservices Backend Engineer,
+49 15757 486 476

2015-09-11 6:30 GMT+02:00 Brian J. Cardiff <notifications@github.com>:

**Comment by waterlink** [comment]
It shows the order of things here: http://play.crystal-lang.org/#/r/etm

--
Best Regards,
Alexey Fedorov,
Sr Ruby, Clojure, Crystal, Golang Developer,
Microservices Backend Engineer,
+49 15757 486 476

2015-09-11 11:04 GMT+02:00 Alex Fedorov <waterlink000@gmail.com>:

**Comment by asterite** [comment]
This seems to be fixed in 0.17.4 (using `type_vars`, the name changed) : https://play.crystal-lang.org/#/r/zze
